strawberry/starlite/controller.py

```diff
@@ -0,0 +1,399 @@
+        return await self._get_context()
+
+    async def get_root_value(self) -> Any:
+        return await self._get_root_value()
+
+
+class BaseContext:
+    def __init__(self):
+        self.request: Optional[Union[Request, WebSocket]] = None
+        self.response: Optional[Response] = None
+
+
+def make_graphql_controller(
+    schema: BaseSchema,
+    path: str = "",
+    websocket_path: str = "/ws",
+    graphiql: bool = True,
+    allow_queries_via_get: bool = True,
+    keep_alive: bool = False,
+    keep_alive_interval: float = 1,
+    debug: bool = False,
+    root_value_getter=None,
+    context_getter=None,
+    subscription_protocols=(GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL),
+    connection_init_wait_timeout: timedelta = timedelta(minutes=1),
+) -> Type[Controller]:
+    routes_path = path
+
+    if not websocket_path:
+        raise ImproperlyConfiguredException(detail="webocket_path must not be empty")
+
+    if context_getter is None:
+
+        def custom_context_getter_():
+            return None
+
+    else:
+        custom_context_getter_ = context_getter
+
+    if root_value_getter is None:
+
+        def root_value_getter_():
+            return None
+
+    else:
+        root_value_getter_ = root_value_getter
+
+    class GraphQLController(Controller):
+        path: str = routes_path
+        dependencies: Optional["Dependencies"] = {
+            "custom_context": Provide(custom_context_getter_),
+            "context": Provide(_context_getter),
+            "root_value": Provide(root_value_getter_),
+        }
+        graphql_ws_handler_class = GraphQLWSHandler
+        graphql_transport_ws_handler_class = GraphQLTransportWSHandler
+
+        _schema: BaseSchema = schema
+        _graphiql: bool = graphiql
+        _allow_queries_via_get: bool = allow_queries_via_get
+        _keep_alive: bool = keep_alive
+        _keep_alive_interval: float = keep_alive_interval
+        _debug: bool = debug
+        _protocols = subscription_protocols
+        _connection_init_wait_timeout: timedelta = connection_init_wait_timeout
+
+        async def execute(
+            self,
+            query: str,
+            variables: Optional[Dict[str, Any]] = None,
+            context: Any = None,
+            operation_name: Optional[str] = None,
+            root_value: Any = None,
+            allowed_operation_types: Optional[Iterable[OperationType]] = None,
+        ):
+            if self._debug:
+                pretty_print_graphql_operation(operation_name, query, variables)
+
+            return await self._schema.execute(
+                query,
+                root_value=root_value,
+                variable_values=variables,
+                operation_name=operation_name,
+                context_value=context,
+                allowed_operation_types=allowed_operation_types,
+            )
+
+        async def process_result(self, result: ExecutionResult) -> GraphQLHTTPResponse:
```

Comment:
Static method. You can also move most of the controller logic and methods into functions. I think it makes for cleaner code, but it's a matter if preference.